pxr/usdImaging/bin/usdrecord/usdrecord.cpp

```diff
@@ -0,0 +1,709 @@
+        // Note that the size of the widget doesn't actually affect the size of
+        // the output image. We just pass it along for cleanliness.
+
+        // Initialize the library
+        if (!glfwInit())
+            return -1;
+
+        glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE);
+
+        // Create a windowed mode window and its OpenGL context
+        GLFWwindow* window = glfwCreateWindow(args.imageWidth, args.imageWidth, "usdrecord", NULL, NULL);
+        if (!window)
+        {
+            glfwTerminate();
+            return -1;
+        }
+
+        // Make the window's context current
+        glfwMakeContextCurrent(window);
+    }
+#endif
+
+    TfToken rendererPluginId = rendererArgs_GetPluginIdFromArgument(args.rendererPlugin);
+
+    // This is the only the necessary part of RefinementComplexities python class needed for usdrecord.
+    static const std::unordered_map<std::string, float> complexities {
+            {"low", 1.0},
+            {"medium", 1.1},
+            {"high", 1.2},
+            {"veryhigh", 1.3}};
+
+    float complexity = complexities.at(args.complexity);
+
+    std::unique_ptr<UsdAppUtilsFrameRecorder> frameRecorder = std::make_unique<UsdAppUtilsFrameRecorder>(rendererPluginId, gpuEnabled);
+
+    if (!rsPrimPath.empty()) {
+        frameRecorder->SetActiveRenderSettingsPrimPath(SdfPath(rsPrimPath));
+    }
+
+    if (!args.rpPrimPath.empty()) {
+        frameRecorder->SetActiveRenderPassPrimPath(SdfPath(args.rpPrimPath));
+    }
+
+    frameRecorder->SetImageWidth(imageWidth);
+    frameRecorder->SetComplexity(complexity);
+    frameRecorder->SetCameraLightEnabled(args.cameraLightEnabled);
+    frameRecorder->SetColorCorrectionMode(colorCorrectionMode);
+    frameRecorder->SetIncludedPurposes(purposes);
+    frameRecorder->SetDomeLightVisibility(args.domeLightVisibility);
+
+    for (const auto &timeCode: frames) {
+        std::cout << "Recording time code: " << timeCode << std::endl;
+
+        std::string outputImagePath = outputImagePathPrefix;
+        if (frameNumberFormater != nullptr) {
+            // if we have a frame number formatter - then that means we successfully parsed a frame number place holder
+            // otherwise the entire original filename is in the prefix string.
+            std::stringstream ss;
+            ss << frameNumberFormater;
+            ss << timeCode << "." << outputImagePathSuffix;
+            outputImagePath += ss.str();
+        }
+
+        try {
+            frameRecorder->Record(usdStage, usdCamera, timeCode, outputImagePath);
+        } catch (TfBaseException e) {
+            std::cerr << "Recording aborted due to the following failure at time code " << timeCode << ": "
+                      << e.what() << std::endl;
+            return 1;
+        }
+    }
+
+    // Release our reference to the frame recorder so it can be deleted before other resources are freed
+    frameRecorder = nullptr;
+
+#if PXR_USE_GLFW
+    glfwTerminate();
```

Comment:
I noticed there's a few return code paths that don't terminate the `glfw` context.

I'm wondering if an RAII pattern could be used here so that the `glfw` context is always terminated.

Reply:
I added a RAII container for the GLFW resources - I'm not totally convinced if this is the best route for usdrecord, if thats the only app that needs to manage this OpenGL context creation in C++.  It's a lot more code to add, instead of just fixing up the missing `glfwTerminate()`.  I do understand the RAII approach being better in general, but here maybe the application is simple enough?  

Anyway, I added it so we can compare approaches, and it's easy enough to revert if thats deemed preferred.  I don't really have strong opinions either way.

Reply:
It might be a reasonable case for a generalized "finally" utility in `Tf`.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c30-define-a-destructor-if-a-class-needs-an-explicit-action-at-object-destruction